src/main/java/uk/gov/pay/directdebit/webhook/gocardless/api/GoCardlessWebhookParser.java

```diff
@@ -0,0 +1,66 @@
+package uk.gov.pay.directdebit.webhook.gocardless.api;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.slf4j.Logger;
+import uk.gov.pay.directdebit.app.logger.PayLoggerFactory;
+import uk.gov.pay.directdebit.payments.model.GoCardlessEvent;
+import uk.gov.pay.directdebit.payments.model.GoCardlessResourceType;
+import uk.gov.pay.directdebit.webhook.gocardless.exception.WebhookParserException;
+
+import javax.inject.Inject;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GoCardlessWebhookParser {
+    private static final Logger LOGGER = PayLoggerFactory.getLogger(GoCardlessWebhookParser.class);
+
+    private ObjectMapper objectMapper;
+
+    @Inject
+    public GoCardlessWebhookParser(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
+    public List<GoCardlessEvent> parse(String webhookPayload) {
+        try {
+            List<GoCardlessEvent> events = new ArrayList<>();
+            JsonNode webhookJson = objectMapper.readTree(webhookPayload);
+            JsonNode eventsPayload = webhookJson.get("events");
+            for (JsonNode eventNode: eventsPayload) {
+                String resourceType = eventNode.get("resource_type").asText();
+                GoCardlessResourceType handledGoCardlessResourceType = GoCardlessResourceType.fromString(resourceType);
+                String resourceId = extractResourceIdFrom(eventNode, handledGoCardlessResourceType);
+                GoCardlessEvent event = new GoCardlessEvent(
+                        null,
+                        eventNode.get("id").asText(),
+                        eventNode.get("action").asText(),
+                        handledGoCardlessResourceType,
+                        eventNode.toString(),
+                        ZonedDateTime.parse(eventNode.get("created_at").asText())
+                );
+                event.setResourceId(resourceId);
+                events.add(event);
+                LOGGER.info("Successfully parsed gocardless webhook, event resource type: {}, action: {}, resource id {}",
+                        event.getResourceType(),
+                        event.getAction(),
+                        event.getResourceId());
+            }
+            return events;
+        } catch (Exception exc) {
+            throw new WebhookParserException("Failed to parse webhooks, body: " + webhookPayload);
+        }
+    }
+
+    private String extractResourceIdFrom(JsonNode jsonNode, GoCardlessResourceType goCardlessResourceType) {
+        switch (goCardlessResourceType) {
+            case PAYMENTS:
+                return jsonNode.get("links").get("payment").asText();
+            case MANDATES:
+                return jsonNode.get("links").get("mandate").asText();
+            default:
+                return null;
```

Comment:
I am not of a big fan of returning optionals for internal method calls. I find them useful for public-facing APIs, of course. Will take a look.

Reply:
Just took a look at the code. I personally don't see any problem calling `event.setResourceId(resourceId);` even when `resourceId` is null. Even if we don't call it, it would be null by default - I think it makes the code simpler. 

Am I missing something? If you have strong opinions about making it clear that `resourceId` could be null, I'll do a null check or return an optional as you suggested :)

Reply:
Yeah, I just thought it might be a bit clearer that we’re only setting it in certain circumstances. Relying on the combination of the facts that `extractResourceIdFrom(JsonNode, GoCardlessResourceType)` can return `null` and that calling `setResourceId(String)` with `null` is an effective no-op feels a little brittle: breaking that coupling (e.g. by returning an `Optional`) and forcing ourselves to deal with that would break the coupling. But I don’t feel too strongly about it :)

Reply:
👍I've implemented that as optional  + `ifPresent(...)`